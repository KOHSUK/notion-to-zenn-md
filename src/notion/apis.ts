import { Client } from '@notionhq/client';
import { Database, Page } from './types';

// Initializing a client
export function getNotionClient(apiSecret: string) {
  const client = new Client({
    auth: apiSecret,
  });

  return client;
}

export async function getDatabase(notion: Client, databaseId: string) {
  const response = await notion.databases.query({
    database_id: databaseId,
  });

  return response.results as Database;
}

export async function getPage(notion: Client, pageId: string) {
  const response = await notion.pages.retrieve({ page_id: pageId });

  return response as Page;
}

export async function getBlocks(notion: Client, pageId: string) {
  const response = await notion.blocks.children.list({
    block_id: pageId,
  });

  return response.results;
}
