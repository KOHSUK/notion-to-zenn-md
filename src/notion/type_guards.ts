import { Bookmark, Breadcrumb, BulletedListItem, Callout, CheckBoxProperty, ChildDatabase, ChildPage, Code, ColumnList, CreatedByProperty, CreatedTimeProperty, DateProperty, Divider, EmailProperty, Embed, Emoji, Equation, File, FilesProperty, FormulaProperty, Heading1, Heading2, Heading3, Image, LastEditedByProperty, LastEditedTimeProperty, LinkPreview, MultiSelectProperty, NumberProperty, NumberedListItem, PDF, Paragraph, PeopleProperty, PhoneNumberProperty, Quote, RelationProperty, RichTextProperty, RollupProperty, SelectProperty, StatusProperty, SyncedBlock, Table, TableOfContents, TableRow, TitleProperty, ToDo, Toggle, UrlProperty, Video } from "./types";

// ##############################################
// Block
// ##############################################

// Type Guards for Bookmark
export function isBookmarkBlock(block: unknown): block is Bookmark {
  return (block as Bookmark).type === "bookmark";
}

// Type Guards for BreadCrumb
export function isBreadcrumbBlock(block: unknown): block is Breadcrumb {
  return (block as Breadcrumb).type === "breadcrumb";
}

// Type Guards for BulletedListItem
export function isBulletedListItemBlock(block: unknown): block is BulletedListItem {
  return (block as BulletedListItem).type === "bulleted_list_item";
}

// Type Guards for Callout
export function isCalloutBlock(block: unknown): block is Callout {
  return (block as Callout).type === "callout";
}

// Type Guards for ChildDatabase
export function isChildDatabaseBlock(block: unknown): block is ChildDatabase {
  return (block as ChildDatabase).type === "child_database";
}

// Type Guards for ChildPage
export function isChildPageBlock(block: unknown): block is ChildPage {
  return (block as ChildPage).type === "child_page";
}

// Type Guards for Code
export function isCodeBlock(block: unknown): block is Code {
  return (block as Code).type === "code";
}

// Type Guards for ColumnList
export function isColumnListBlock(block: unknown): block is ColumnList {
  return (block as ColumnList).type === "column_list";
}

// Type Guards for Divider
export function isDividerBlock(block: unknown): block is Divider {
  return (block as Divider).type === "divider";
}

// Type Guards for Embed
export function isEmbedBlock(block: unknown): block is Embed {
  return (block as Embed).type === "embed";
}

// Type Guards for Equation
export function isEquationBlock(block: unknown): block is Equation {
  return (block as Equation).type === "equation";
}

// Type Guards for File
export function isFileBlock(block: unknown): block is File {
  return (block as File).type === "file";
}

// Type Guards for Heading1
export function isHeading1Block(block: unknown): block is Heading1 {
  return (block as Heading1).type === "heading_1";
}

// Type Guards for Heading2
export function isHeading2Block(block: unknown): block is Heading2 {
  return (block as Heading2).type === "heading_2";
}

// Type Guards for Heading3
export function isHeading3Block(block: unknown): block is Heading3 {
  return (block as Heading3).type === "heading_3";
}

// Type Guards for Image
export function isImageBlock(block: unknown): block is Image {
  return (block as Image).type === "image";
}

// Type Guards for LinkPreview
export function isLinkPreviewBlock(block: unknown): block is LinkPreview {
  return (block as LinkPreview).type === "link_preview";
}

// Type Guards for NumberedListItem
export function isNumberedListItemBlock(block: unknown): block is NumberedListItem {
  return (block as NumberedListItem).type === "numbered_list_item";
}

// Type Guards for Paragraph
export function isParagraphBlock(block: unknown): block is Paragraph {
  return (block as Paragraph).type === "paragraph";
}

// Type Guards for PDF
export function isPDFBlock(block: unknown): block is PDF {
  return (block as PDF).type === "pdf";
}

// Type Guards for Quote
export function isQuoteBlock(block: unknown): block is Quote {
  return (block as Quote).type === "quote";
}

// Type Guards for SyncedBlock
export function isSyncedBlockBlock(block: unknown): block is SyncedBlock {
  return (block as SyncedBlock).type === "synced_block";
}

// Type Guards for Table
export function isTableBlock(block: unknown): block is Table {
  return (block as Table).type === "table";
}

// Type Guards for TableRow
export function isTableRowBlock(block: unknown): block is TableRow {
  return (block as TableRow).type === "table_row";
}

// Type Guards for TableOfContents
export function isTableOfContentsBlock(block: unknown): block is TableOfContents {
  return (block as TableOfContents).type === "table_of_contents";
}

// Type Guards for ToDo
export function isToDoBlock(block: unknown): block is ToDo {
  return (block as ToDo).type === "to_do";
}

// Type Guards for Toggle
export function isToggleBlock(block: unknown): block is Toggle {
  return (block as Toggle).type === "toggle";
}

// Type Guards for Video
export function isVideoBlock(block: unknown): block is Video {
  return (block as Video).type === "video";
}

// ##############################################
// Property
// ##############################################

// Type Guards for CheckBox

export function isCheckBoxProperty(property: unknown): property is CheckBoxProperty {
  return (property as CheckBoxProperty).type === "checkbox";
}

// Type Guards for CreatedByProperty
export function isCreatedByProperty(property: unknown): property is CreatedByProperty {
  return (property as CreatedByProperty).type === "created_by";
}

// Type Guards for CreatedTimeProperty
export function isCreatedTimeProperty(property: unknown): property is CreatedTimeProperty {
  return (property as CreatedTimeProperty).type === "created_time";
}

// Type Guards for DateProperty
export function isDateProperty(property: unknown): property is DateProperty {
  return (property as DateProperty).type === "date";
}

// Type Guards for EmailProperty
export function isEmailProperty(property: unknown): property is EmailProperty {
  return (property as EmailProperty).type === "email";
}

// Type Guards for FilesProperty
export function isFilesProperty(property: unknown): property is FilesProperty {
  return (property as FilesProperty).type === "files";
}

// Type Guards for FormulaProperty
export function isFormulaProperty(property: unknown): property is FormulaProperty {
  return (property as FormulaProperty).type === "formula";
}

// Type Guards for LastEditedByProperty
export function isLastEditedByProperty(property: unknown): property is LastEditedByProperty {
  return (property as LastEditedByProperty).type === "last_edited_by";
}

// Type Guards for LastEditedTimeProperty
export function isLastEditedTimeProperty(property: unknown): property is LastEditedTimeProperty {
  return (property as LastEditedTimeProperty).type === "last_edited_time";
}

// Type Guards for MultiSelectProperty
export function isMultiSelectProperty(property: unknown): property is MultiSelectProperty {
  return (property as MultiSelectProperty).type === "multi_select";
}

// Type Guards for NumberProperty
export function isNumberProperty(property: unknown): property is NumberProperty {
  return (property as NumberProperty).type === "number";
}

// Type Guards for PeopleProperty
export function isPeopleProperty(property: unknown): property is PeopleProperty {
  return (property as PeopleProperty).type === "people";
}

// Type Guards for PhoneNumberProperty
export function isPhoneNumberProperty(property: unknown): property is PhoneNumberProperty {
  return (property as PhoneNumberProperty).type === "phone_number";
}

// Type Guards for RelationProperty
export function isRelationProperty(property: unknown): property is RelationProperty {
  return (property as RelationProperty).type === "relation";
}

// Type Guards for RichTextProperty
export function isRichTextProperty(property: unknown): property is RichTextProperty {
  return (property as RichTextProperty).type === "rich_text";
}

// Type Guards for RollupProperty
export function isRollupProperty(property: unknown): property is RollupProperty {
  // Note: The type string for RollupProperty was not provided.
  return (property as RollupProperty).type === "rollup";
}

// Type Guards for SelectProperty
export function isSelectProperty(property: unknown): property is SelectProperty {
  return (property as SelectProperty).type === "select";
}

// Type Guards for StatusProperty
export function isStatusProperty(property: unknown): property is StatusProperty {
  return (property as StatusProperty).type === "status";
}

// Type Guards for TitleProperty
export function isTitleProperty(property: unknown): property is TitleProperty {
  return (property as TitleProperty).type === "title";
}

// Type Guards for UrlProperty
export function isUrlProperty(property: unknown): property is UrlProperty {
  return (property as UrlProperty).type === "url";
}

export function isEmojiProperty(property: unknown): property is Emoji {
  return (property as Emoji).type === "emoji";
}