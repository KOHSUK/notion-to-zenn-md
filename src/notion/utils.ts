import { RichText, Title } from "./types";

export function getRichTextString(richText?: RichText[]): string {
  if (!richText || richText.length < 1) {
    return "";
  }

  return richText[0].plain_text || "";
}

export function getTitleString(title?: Title): string {
  if (!title || title.length < 1) {
    return "";
  }

  return title[0].plain_text || "";
}

