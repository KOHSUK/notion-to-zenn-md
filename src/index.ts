import { NotionToMarkdown } from "notion-to-md";
import { NotionClient, getNotionClient, getPage } from "./notion";
import { getTransformer } from "./transformer";
import { FrontMatter } from "./zenn";
import { getTitleString } from "./notion/utils";
import { isCheckBoxProperty, isDateProperty, isEmojiProperty, isMultiSelectProperty, isSelectProperty, isTitleProperty } from "./notion/type_guards";
import { decodeUnicodeEscapeSequence } from "./utils";
import { format, parseISO } from "date-fns";

export type FrontMatterNotionPropMapping = {
  [key in keyof Omit<FrontMatter, 'emoji'>]: string;
}

const defaultMapping: FrontMatterNotionPropMapping = {
  title: 'Title',
  type: 'Type',
  topics: 'Topics',
  published: 'Published',
} as const;

export default class NotionToZennMarkdown {
  private transformer: NotionToMarkdown;
  private client: NotionClient;

  constructor(apiSecret: string) {
    this.client = getNotionClient(apiSecret);
    this.transformer = getTransformer(this.client);
  }

  /**
   * 指定したNotionのページIDのページをZenn記法のマークダウン(FrontMatterなし)に変換して返します。
   * @param pageId Notion Page ID
   * @returns Markdown representation of the page
   */
  async pageToZennMarkdown(pageId: string): Promise<string> {
    const mdBlocks = await this.transformer.pageToMarkdown(pageId);
    const mdString = this.transformer.toMarkdownString(mdBlocks);

    return mdString.parent;
  }

  /**
   * 指定したNotionのページIDのページプロパティからZenn記法のマークダウンのFrontMatterオブジェクトを生成して返します。
   * @param pageId Notion Page ID
   * @param mappingKeys Mapping Keys to Zenn Markdown Front Matter Keys
   * @returns Front Matter
   */
  async getFrontMatter(pageId: string, mappingKeys: Partial<FrontMatterNotionPropMapping> = defaultMapping): Promise<FrontMatter> {
    const mapping = { ...defaultMapping, ...mappingKeys };
    const page = await getPage(this.client, pageId);

    if (!page) {
      throw new Error(`Page not found. pageId: ${pageId}`);
    }

    // title
    const titleProp = page.properties[mapping.title];
    let title = "";
    if (titleProp && isTitleProperty(titleProp)) {
      title = getTitleString(titleProp.title)
    }

    const emojiProp = page.icon
    let emoji = "";
    if (emojiProp && isEmojiProperty(emojiProp)) {
      emoji = emojiProp.emoji;
    }

    const typeProp = page.properties[mapping.type];
    let type = "";
    if (typeProp && isSelectProperty(typeProp)) {
      type = typeProp.select.name;
    }

    const topicsProp = page.properties[mapping.topics];
    let topics: string[] = [];
    if (topicsProp && isMultiSelectProperty(topicsProp)) {
      topics = topicsProp.multi_select.map((val) => val.name);
    }

    const publishedProp = page.properties[mapping.published];
    let published = false;
    if (publishedProp && isCheckBoxProperty(publishedProp)) {
      published = publishedProp.checkbox;
    }

    let publishedAt = undefined;
    if (mapping.publishedAt) {
      const publishedAtProp = page.properties[mapping.publishedAt];
      if (publishedAtProp && isDateProperty(publishedAtProp)) {
        publishedAt = format(parseISO(publishedAtProp.date.start), 'yyyy-MM-dd HH:mm');
      }
    }

    const frontMatter: FrontMatter = {
      title,
      emoji,
      type,
      topics,
      published,
      publishedAt
    }

    return frontMatter;
  }

  /**
   * マークダウン文字列を生成します。
   * @param mdString マークダウンの文字列
   * @param frontMatter Front Matterのオブジェクト
   * @returns マークダウンの文字列
   */
  private _generateMd(mdString: string, frontMatter: FrontMatter): string {
    // // topicsをカスタムフォーマットに変換
    // const topicPlaceholder = "__TOPIC_PLACEHOLDER__";
    // const customTopics = `[${frontMatter.topics.map(topic => `"${topic}"`).join(', ')}]`;
    // // emojiを変換
    // const emojiPlaceholder = "__EMOJI_PLACEHOLDER__";
    // const decodedEmoji = decodeUnicodeEscapeSequence(frontMatter.emoji);
    // const modifiedFrontMatter = { ...frontMatter, topics: topicPlaceholder, emoji: emojiPlaceholder };
    // const md = matter.stringify(mdString, modifiedFrontMatter).replace(topicPlaceholder, customTopics).replace(emojiPlaceholder, decodedEmoji);

    // return md
    // topicsをカスタムフォーマットに変換
    const title = `title: "${frontMatter.title}"`;
    const emoji = `emoji: "${decodeUnicodeEscapeSequence(frontMatter.emoji)}"`;
    const type = `type: "${frontMatter.type}"`;
    const topics = `topics: [${frontMatter.topics.map(topic => `"${topic}"`).join(', ')}]`;
    const published = `published: ${frontMatter.published}`
    const publishedAt = frontMatter.publishedAt ? `published_at: ${frontMatter.publishedAt}` : "";

    return `---\n${title}\n${emoji}\n${type}\n${topics}\n${published}${publishedAt ? `\n${publishedAt}` : ""}\n---${mdString ? `\n${mdString}` : ""}`;
  }

  /**
   * 指定したNotionのページIDのページプロパティからFrontMatter文字列を生成します。
   * @param pageId Notion Page ID
   * @param mappingKeys Mapping Keys to Zenn Markdown Front Matter Keys
   * @returns FrontMatterの文字列を生成
   */
  async getFrontMatterString(pageId: string, mappingKeys?: Partial<FrontMatterNotionPropMapping>): Promise<string> {
    const frontMatter = await this.getFrontMatter(pageId, mappingKeys);
    const frontMatterString = this._generateMd("", frontMatter);

    return frontMatterString;
  }

  /**
   * マークダウン文字列を生成します。
   * @param mdString マークダウンの文字列
   * @param frontMatter Front Matterのオブジェクト
   * @returns マークダウンの文字列
   */
  stringify(mdString: string, fontMatter: FrontMatter): string {
    const frontMatterString = this._generateMd(mdString, fontMatter);

    return frontMatterString;
  }

  /**
   * 指定したNotionのページIDのページをZenn記法のマークダウンに変換して返します。
   * @param pageId Notion Page ID
   * @param mappingKeys Mapping Keys to Zenn Markdown Front Matter Keys
   * @returns Markdown representation of the page
   */
  async generateMd(pageId: string, mappingKeys?: Partial<FrontMatterNotionPropMapping>): Promise<string> {
    const mdString = await this.pageToZennMarkdown(pageId);
    const frontMatter = await this.getFrontMatter(pageId, mappingKeys);
    const zennMarkdown = this._generateMd(mdString, frontMatter);

    return zennMarkdown;
  }

  /**
   * Markdown文字列から画像のURLのリストを抽出します。
   * @param mdString Markdown文字列
   * @returns Image paths
   */
  extractImageUrls(mdString: string): string[] {
    const imagePathRegex = /!\[.*?\]\((http[s]?:\/\/[^\s)]+|[^)\s]+)\)/g;
    const matches = mdString.match(imagePathRegex);

    if (!matches) {
      return [];
    }

    const imagePaths = matches.map((match) => {
      const pathMatch = match.match(/!\[.*?\]\((http[s]?:\/\/[^\s)]+|[^)\s]+)\)/);
      if (pathMatch && pathMatch[1]) {
        return pathMatch[1];
      }
      return "";
    });

    return imagePaths.filter((path) => path !== "");
  }

  /**
   * 指定したNotionのページIDのページに含まれる画像のURLのリストを返します。
   * @param pageId Notion Page ID
   * @returns Image paths
   */
  async listImageUrls(pageId: string): Promise<string[]> {
    const markdown = await this.pageToZennMarkdown(pageId);
    const imageUrls = this.extractImageUrls(markdown);

    return imageUrls;
  }
}