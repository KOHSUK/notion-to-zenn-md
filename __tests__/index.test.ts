// test index.ts
import NotionToZennMarkdown from "../src/index";
import { describe, it, vi, expect, beforeAll } from 'vitest';
import { NotionToMarkdown } from 'notion-to-md';
import * as notion from '../src/notion';
import { Page } from "../src/notion";


describe("NotionToZennMarkdown", () => {
  describe('generateMd', () => {

    beforeAll(() => {
      const mockPageToMarkdown = vi.spyOn(NotionToMarkdown.prototype, "pageToMarkdown")

      mockPageToMarkdown.mockImplementation(async () => {
        return [
          {
            type: 'heading_1',
            blockId: '70ed00de-6ff2-43a5-be32-a8991b25f1ca',
            parent: '# Heading 1',
            children: []
          },
          {
            type: 'heading_2',
            blockId: '3b8bed17-94a3-452d-84bc-274508ee6019',
            parent: '## Heading 2',
            children: []
          },
          {
            type: 'heading_3',
            blockId: 'd1941f4a-0d86-41ec-93c1-ed9bd2aa749d',
            parent: '### Heading 3',
            children: []
          },
          {
            type: 'paragraph',
            blockId: 'e0c2eb47-d4eb-45f7-abc3-004bb369d113',
            parent: 'Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec interdum eros a metus imperdiet maximus. Aenean gravida lacinia ultricies. Duis sit amet justo et lectus pharetra accumsan. Pellentesque urna erat.(text)',
            children: []
          },
          {
            type: 'paragraph',
            blockId: '34abf6cd-013f-4f29-ae2e-6d415bcc02cb',
            parent: '_イタリック_',
            children: []
          },
          {
            type: 'paragraph',
            blockId: '4f1c4436-d67d-47a2-a0c6-da6ad46b7359',
            parent: '**太字**',
            children: []
          },
          {
            type: 'paragraph',
            blockId: 'e35d4400-f27a-4017-828d-a1ebbbb87e6f',
            parent: '~~取り消し線~~',
            children: []
          },
          {
            type: 'paragraph',
            blockId: '51cc934b-4c12-46f6-8ec9-3fdc1afd760b',
            parent: 'インラインの`code`の挿入',
            children: []
          },
          {
            type: 'paragraph',
            blockId: '3fc190f1-9a12-405a-89f5-8c456836cac0',
            parent: '[テキストのリンク](https://www.kohsuk.tech/articles/blog-migration-to-nextjs-and-notion)',
            children: []
          },
          {
            type: 'to_do',
            blockId: 'b8ed1cfc-2c66-4b48-b644-b2769bf4e3c5',
            parent: '- [ ] CheckBox(off)',
            children: []
          },
          {
            type: 'to_do',
            blockId: 'ddf19014-f707-4ece-842b-8651dc9f8a4d',
            parent: '- [x] CheckBox(on)',
            children: []
          },
          {
            type: 'bulleted_list_item',
            blockId: 'aecf9141-13e3-4967-a81d-64cb99611dbb',
            parent: '- Bulleted List 1',
            children: []
          },
          {
            type: 'bulleted_list_item',
            blockId: 'ec2e794c-fb2d-4326-929b-6cf22688ce3c',
            parent: '- Bulleted List 2',
            children: []
          },
          {
            type: 'numbered_list_item',
            blockId: '9d8f9045-f685-4ea3-a53a-9ece5857ca2f',
            parent: '1. Numbered List 1',
            children: []
          },
          {
            type: 'numbered_list_item',
            blockId: '6e11cfa2-7fad-4d2f-bbb6-a771cb220443',
            parent: '2. Numbered List 2',
            children: []
          },
          {
            type: 'toggle',
            blockId: 'b19aab0f-9a7c-4eb6-80a3-21ec0570d9bc',
            parent: 'Toggle List',
            children: []
          },
          {
            type: 'quote',
            blockId: '3ccd1f97-eaa5-4c61-86bb-6098307a8725',
            parent: '> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec interdum eros a metus imperdiet maximus. Aenean gravida lacinia ultricies. Duis sit amet justo et lectus pharetra accumsan. Pellentesque urna erat.(Quote)',
            children: []
          },
          {
            type: 'divider',
            blockId: '7ace8a6b-12cb-427d-b238-24d869f0d2c9',
            parent: '---',
            children: []
          },
          {
            type: 'callout',
            blockId: 'e5492ec7-173a-439b-a2de-7b7b8ff17a0c',
            parent: '\n:::message\n💡 Callout\n:::\n',
            children: []
          },
          {
            type: 'image',
            blockId: 'f55c3a72-5797-44a5-a5f7-73615c0784e4',
            parent: '![画像（キャプションの表示）](https://prod-files-secure.s3.us-west-2.amazonaws.com/87fa912e-5725-43e4-91de-2911f2fd0b15/49c4adce-aafb-47cb-ba90-a4b31e4d5a3e/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=XXXXXXXXXXXXXXXXXXXX%2F20231006%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20231006T110453Z&X-Amz-Expires=3600&X-Amz-Signature=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&X-Amz-SignedHeaders=host&x-id=GetObject)\n' +
              '*画像（キャプションの表示）*',
            children: []
          },
          {
            type: 'bookmark',
            blockId: 'a644aa46-46ba-45d0-af8e-49ee403d431f',
            parent: '\n' +
              '@[card](https://www.kohsuk.tech/articles/blog-migration-to-nextjs-and-notion)\n',
            children: []
          },
          {
            type: 'paragraph',
            blockId: '1b4b4371-afdd-4aae-8747-f64a8dd30422',
            parent: '',
            children: []
          },
          {
            type: 'paragraph',
            blockId: '5b8fa501-05ae-4dd3-9f11-fc19a6b16ac6',
            parent: '',
            children: []
          }
        ];

      });

      const getPageMock = vi.spyOn(notion, 'getPage');

      getPageMock.mockImplementation(async () => ({
        "object": "page",
        "id": "99999999-9999-9999-9999-999999999991",
        "created_time": "2023-10-01T13:52:00.000Z",
        "last_edited_time": "2023-10-01T14:30:00.000Z",
        "created_by": {
          "object": "user",
          "id": "99999999-9999-9999-9999-999999999999"
        },
        "last_edited_by": {
          "object": "user",
          "id": "99999999-9999-9999-9999-999999999999"
        },
        "cover": null,
        "icon": { "type": "emoji", "emoji": "🤩" },
        "parent": {
          "type": "database_id",
          "database_id": "b9121b4b-2e23-4370-815f-acbd5b6d50da"
        },
        "archived": false,
        "properties": {
          "Type": {
            "id": "%3D%5BO%3C",
            "type": "select",
            "select": { "id": "]ns^", "name": "tech", "color": "blue" }
          },
          "Published": { "id": "H%7BJ%7C", "type": "checkbox", "checkbox": true },
          "Topics": {
            "id": "ZzyO",
            "type": "multi_select",
            "multi_select": [
              {
                "id": "c1e9d580-ae07-4679-a879-4d97c9d7e3e3",
                "name": "notion",
                "color": "brown"
              }
            ]
          },
          "Title": {
            "id": "title",
            "type": "title",
            "title": [
              {
                "type": "text",
                "text": { "content": "記事のタイトルです。", "link": null },
                "annotations": {
                  "bold": false,
                  "italic": false,
                  "strikethrough": false,
                  "underline": false,
                  "code": false,
                  "color": "default"
                },
                "plain_text": "記事のタイトルです。",
                "href": null
              }
            ]
          },
          "CustomType": {
            "id": "%3D%5BO%3C",
            "type": "select",
            "select": { "id": "]ns^", "name": "tech", "color": "blue" }
          },
          "CustomPublished": { "id": "H%7BJ%7C", "type": "checkbox", "checkbox": true },
          "CustomTopics": {
            "id": "ZzyO",
            "type": "multi_select",
            "multi_select": [
              {
                "id": "c1e9d580-ae07-4679-a879-4d97c9d7e3e3",
                "name": "notion",
                "color": "brown"
              }
            ]
          },
          "CustomPublishedAt": {
            "id": "xxxxxxxx",
            "type": "date",
            "date": { "start": "2023-10-01T13:52:00.000Z", "end": null, "time_zone": null }
          }
        },
        "url": "https://www.notion.so/99999999999999999999999999999991",
        "public_url": null
      } as Page));

    })

    it('should generate markdown properly', async () => {
      const n2zm = new NotionToZennMarkdown('sample');
      const md = await n2zm.generateMd('');

      expect(md).toMatchSnapshot();
    });

    it('should generate markdown properly even if a mapping key is specified', async () => {
      const n2zm = new NotionToZennMarkdown('secret_sample');
      const md = await n2zm.generateMd('page_id', {
        type: 'CustomType',
        published: 'CustomPublished',
        topics: 'CustomTopics'
      });

      expect(md).toMatchSnapshot();
    });
  });

  describe('getFrontMatterString', () => {
    it('should output published_at if published_at is specified in a Mapping Key', async () => {
      const n2zm = new NotionToZennMarkdown('secret_sample');
      const frontMatter = await n2zm.getFrontMatterString('page_id', {
        publishedAt: 'CustomPublishedAt'
      });

      expect(frontMatter).toBe(`---
title: "記事のタイトルです。"
emoji: "🤩"
type: "tech"
topics: ["notion"]
published: true
published_at: 2023-10-01 13:52
---`);
    });

    it('should not output published_at if published_at is not specified in a Mapping Key', async () => {
      const n2zm = new NotionToZennMarkdown('secret_sample');
      const frontMatter = await n2zm.getFrontMatterString('page_id');

      expect(frontMatter).toBe(`---
title: "記事のタイトルです。"
emoji: "🤩"
type: "tech"
topics: ["notion"]
published: true
---`);
    });

  })

  describe('extractImageUrls', () => {
    const markdown = `
# My Favorite Places to Visit

I love to travel and explore new places. Here are some of my favorite destinations:

## Paris, France

![Eiffel Tower](https://via.placeholder.com/500x300)

Paris is known as the City of Love, and it's easy to see why. The architecture, the food, and the culture all combine to create a romantic atmosphere that is hard to resist. One of my favorite things to do in Paris is to visit the Eiffel Tower. The view from the top is breathtaking.

## Bali, Indonesia

![Bali Beach](https://via.placeholder.com/500x400)

Bali is a tropical paradise that is perfect for relaxation and adventure. The beaches are beautiful, and the water is warm and clear. I love to go surfing in Bali, and there are plenty of great spots for beginners and experts alike.

## Tokyo, Japan

![Tokyo Skyline](https://via.placeholder.com/400x300)

Tokyo is a city that never sleeps. There is always something to do or see, from visiting the temples and shrines to exploring the bustling streets and markets. One of my favorite things to do in Tokyo is to visit the Tsukiji Fish Market. The sushi there is some of the best I've ever had.

## New York City, USA

![New York City Skyline](https://via.placeholder.com/500x200)

New York City is the city that never sleeps. There is always something to do or see, from visiting the museums and galleries to exploring the parks and neighborhoods. One of my favorite things to do in New York City is to visit Central Park. It's a great place to relax and enjoy nature in the middle of the city.

Overall, these are just a few of my favorite places to visit. Each destination has its own unique charm and appeal, and I can't wait to explore more of the world in the future.
`
    it('should extract image urls', () => {
      const n2zm = new NotionToZennMarkdown('sample');

      const imageUrls = n2zm.extractImageUrls(markdown);

      expect(imageUrls).toEqual([
        'https://via.placeholder.com/500x300',
        'https://via.placeholder.com/500x400',
        'https://via.placeholder.com/400x300',
        'https://via.placeholder.com/500x200'
      ]);
    });

    it('should return empty array if there is no image', () => {
      const n2zm = new NotionToZennMarkdown('sample');

      const imageUrls = n2zm.extractImageUrls('');

      expect(imageUrls).toEqual([]);
    });

  });

  describe('listImageUrls', () => {
    it('should list image urls', async () => {
      const n2zm = new NotionToZennMarkdown('sample');

      const imageUrls = await n2zm.listImageUrls('page_id');

      expect(imageUrls).toEqual([
        'https://prod-files-secure.s3.us-west-2.amazonaws.com/87fa912e-5725-43e4-91de-2911f2fd0b15/49c4adce-aafb-47cb-ba90-a4b31e4d5a3e/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=XXXXXXXXXXXXXXXXXXXX%2F20231006%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20231006T110453Z&X-Amz-Expires=3600&X-Amz-Signature=XXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX&X-Amz-SignedHeaders=host&x-id=GetObject'
      ]);
    });
  })

});


