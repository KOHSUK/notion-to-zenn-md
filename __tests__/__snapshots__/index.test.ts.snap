// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`NotionToZennMarkdown >  1`] = `
"---
title: \\"記事のタイトルです。\\"
emoji: \\"🤩\\"
type: \\"tech\\"
topics: [\\"notion\\"]
published: true
---

# Heading 1


## Heading 2


### Heading 3


Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec interdum eros a metus imperdiet maximus. Aenean gravida lacinia ultricies. Duis sit amet justo et lectus pharetra accumsan. Pellentesque urna erat.(text)


_イタリック_


**太字**


~~取り消し線~~


インラインの\`code\`の挿入


[テキストのリンク](https://www.kohsuk.tech/articles/blog-migration-to-nextjs-and-notion)

- [ ] CheckBox(off)
- [x] CheckBox(on)
- Bulleted List 1
- Bulleted List 2
1. Numbered List 1
2. Numbered List 2

> Lorem ipsum dolor sit amet, consectetur adipiscing elit. Donec interdum eros a metus imperdiet maximus. Aenean gravida lacinia ultricies. Duis sit amet justo et lectus pharetra accumsan. Pellentesque urna erat.(Quote)


---



:::message
💡 Callout
:::



![画像（キャプションの表示）](https://prod-files-secure.s3.us-west-2.amazonaws.com/87fa912e-5725-43e4-91de-2911f2fd0b15/49c4adce-aafb-47cb-ba90-a4b31e4d5a3e/Untitled.png?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Credential=AKIAT73L2G45EIPT3X45%2F20231006%2Fus-west-2%2Fs3%2Faws4_request&X-Amz-Date=20231006T110453Z&X-Amz-Expires=3600&X-Amz-Signature=60ebe5069c0ecd65a09c708136fd3dd385fe74d619df45229a2efaf2e9385254&X-Amz-SignedHeaders=host&x-id=GetObject)
*画像（キャプションの表示）*



@[card](https://www.kohsuk.tech/articles/blog-migration-to-nextjs-and-notion)


"
`;
